locallib.php

```diff
@@ -346,24 +359,112 @@ function zoom_get_sessions_for_display($meetingid) {
+            // Use 0 as there isn't anything better to return.
+            $cachednextoccurrence = 0;
+
+            // Otherwise we have a recurring meeting with a recurrence schedule.
+        } else {
+            // Get the calendar event of the next occurrence.
+            $selectclause = "modulename = :modulename AND instance = :instance AND (timestart + timeduration) >= :now";
+            $selectparams = array('modulename' => 'zoom', 'instance' => $zoom->id, 'now' => time());
+            $nextoccurence = $DB->get_records_select('event', $selectclause, $selectparams, 'timestart ASC', 'timestart', 0, 1);
+
+            // If we haven't got a single event.
+            if (count($nextoccurence) != 1) {
+                // Use 0 as there isn't anything better to return.
+                $cachednextoccurrence = 0;
+            }
+
+            // Use the timestamp of the event.
+            $nextoccurenceobject = reset($nextoccurence);
+            $cachednextoccurrence = $nextoccurenceobject->timestart;
+        }
+
+        // Store the next occurrence into the cache.
+        $cache->set($zoom->id, $cachednextoccurrence);
+    }
+
+    // Return the next occurrence.
+    return $cachednextoccurrence;
+}
+
 /**
  * Determine if a zoom meeting is in progress, is available, and/or is finished.
  *
  * @param stdClass $zoom
  * @return array Array of booleans: [in progress, available, finished].
  */
 function zoom_get_state($zoom) {
+    // Get plugin config.
     $config = get_config('zoom');
+
+    // Get the current time as calculation basis.
     $now = time();
 
-    $firstavailable = $zoom->start_time - ($config->firstabletojoin * 60);
-    $lastavailable = $zoom->start_time + $zoom->duration;
+    // If this is a recurring meeting with a recurrence schedule.
+    if ($zoom->recurring && $zoom->recurrence_type != ZOOM_RECURRINGTYPE_NOTIME) {
+        // Get the next occurrence time.
+        $nextoccurrence = zoom_get_next_occurrence($zoom);
+
+        // Calculate the time when the recurring meeting becomes available next,
+        // based on the next occurrence start time and the general meeting lead time.
+        $firstavailable = $nextoccurrence - ($config->firstabletojoin * 60);
+
+        // Calculate the time when the meeting ends to be available,
+        // based on the next occurrence start time and the meeting duration.
+        $lastavailable = $nextoccurrence + $zoom->duration;
+
+        // Otherwise.
+    } else {
+        // Calculate the time when the meeting becomes available at first,
+        // based on the meeting start time and the general meeting lead time.
+        $firstavailable = $zoom->start_time - ($config->firstabletojoin * 60);
+
+        // Calculate the time when the meeting ends to be available,
+        // based on the meeting start time and the meeting duration.
+        $lastavailable = $zoom->start_time + $zoom->duration;
+    }
```

Comment:
Prefer to minimize the number of times that duplicate logic occurs. Please consider this:
```suggestion
        // Get the next occurrence time.
        $starttime = zoom_get_next_occurrence($zoom);
    } else {
        $starttime = $zoom->start_time;
    }

    // Calculate the time when the recurring meeting becomes available next,
    // based on the next occurrence start time and the general meeting lead time.
    $firstavailable = $starttime - ($config->firstabletojoin * 60);

    // Calculate the time when the meeting ends to be available,
    // based on the next occurrence start time and the meeting duration.
    $lastavailable = $starttime + $zoom->duration;
```

Reply:
To me it looks logical enough to change, however I'll get @abias (as its his code) to comment if its valid enough to change?

Reply:
Just looking at what the code does, this change is fine for me.

But please note the difference in the comments between line 442-447 and line 445-450: In the first block, we are dealing with a recurring meeting and the next occurrence start time. In the second block, we are dealing with a non-recurring meeting and its start time.

If you do this change, you should adapt the comments to reflect this difference as well.

Reply:
Thanks @abias ! I did not looks closely enough at the comments. I've updated the suggestion with language that I believe addresses both situations.

Reply:
That's fine, @jrchamp !

I would just like to propose to be slightly more verbose in the if-else sections:

```suggestion
        // Get the next occurrence start time.
        $starttime = zoom_get_next_occurrence($zoom);
    } else {
        // Get the meeting start time.
        $starttime = $zoom-&gt;start_time;
    }
          
    // Calculate the time when the meeting will be available to join,
    // subtracting the meeting lead time from the start time.
    $firstavailable = $starttime - ($config-&gt;firstabletojoin * 60);
          
    // Calculate the time when the meeting will stop being available,
    // adding the start time and the meeting duration.
    $lastavailable = $starttime + $zoom->duration;
```

If you agree, @jwalits could merge this change into his working branch from my point of view.

Reply:
Does that look suitable to you @jrchamp? Happy to update the code in my working branch if both of you are happy?

Reply:
Yes, that sounds good. On my screen, it looks like some `&gt;` snuck in where `>` should be in Alex's suggestion, but otherwise, I'm very happy.

Reply:
> On my screen, it looks like some &gt; snuck in where > should be in Alex's suggestion, but otherwise

Oh, you are right. That seems to be some copy & paste issue. I updated my comment above.

@jwalits, could you please proceed with this comment here then?

Reply:
@jwalits, regardless of the other open discussions here, could you please push this small change to the PR so that this discussion could be resolved?

Reply:
I have now updated this in the latest commit